ostd/src/mm/page_table/cursor.rs

```diff
@@ -785,91 +780,45 @@ where
         assert!(src_end <= src.0.barrier_va.end);
 
         while self.0.va < this_end && src.0.va < src_end {
-            let cur_pte = src.0.read_cur_pte();
-            if !cur_pte.is_present() {
-                src.0.move_forward();
-                continue;
-            }
-
-            // Go down if it's not a last node.
-            if !cur_pte.is_last(src.0.level) {
-                src.0.level_down();
-
-                // We have got down a level. If there's no mapped PTEs in
-                // the current node, we can go back and skip to save time.
-                if src.0.guards[(src.0.level - 1) as usize]
-                    .as_ref()
-                    .unwrap()
-                    .nr_children()
-                    == 0
-                {
-                    src.0.level_up();
+            let src_va = src.0.va;
+            let mut src_entry = src.0.cur_entry();
+
+            match src_entry.to_owned() {
+                Child::PageTable(pt) => {
+                    let pt = pt.lock();
+                    // If there's no mapped PTEs in the next level, we can
+                    // skip to save time.
+                    if pt.nr_children() != 0 {
+                        src.0.push_level(pt);
+                    } else {
+                        src.0.move_forward();
+                    }
+                    continue;
+                }
+                Child::None => {
+                    src.0.move_forward();
+                    continue;
+                }
+                Child::Untracked(_, _, _) => {
+                    panic!("Copying untracked mappings");
+                }
+                Child::Page(page, mut prop) => {
+                    // Do protection.
+                    src_entry.protect(op);
+
+                    // Do copy.
+                    op(&mut prop);
+                    self.jump(src_va).unwrap();
+                    let mapped_page_size = page.size();
+                    let original = self.map(page, prop);
+                    assert!(original.is_none());
+
+                    // Only move the source cursor forward since `Self::map` will do it.
+                    // This assertion is to ensure that they move by the same length.
+                    debug_assert_eq!(mapped_page_size, page_size::<C>(src.0.level));
                     src.0.move_forward();
```

Comment:
This might be a slightly better way to group the statements.
```rust
                    let mapped_page_size = page.size();
                    
                    // Do protection.
                    src_entry.protect(op);
                    op(&mut prop);

                    // Do copy.
                    self.jump(src_va).unwrap();
                    let original = self.map(page, prop);
                    assert!(original.is_none());

                    // Only move the source cursor forward since `Self::map` will do it.
                    // This assertion is to ensure that they move by the same length.
                    debug_assert_eq!(mapped_page_size, page_size::<C>(src.0.level));
                    src.0.move_forward();
```